c++/src/kj/compat/http-test.c++

```diff
@@ -6188,6 +6213,91 @@ KJ_TEST("CONNECT Server cancels read w/client") {
+      : fulfiller(kj::mv(fulfiller)) {}
+
+  void setPromise(kj::Promise<void> promise) {
+    task = promise.eagerlyEvaluate([this](kj::Exception&& exception) {
+      if (fulfiller->isWaiting()) {
+        fulfiller->reject(kj::mv(exception));
+      } else {
+        kj::throwRecoverableException(kj::mv(exception));
+      }
+    });
+  }
+
+  kj::Own<kj::AsyncIoStream> accept(
+      uint statusCode, kj::StringPtr statusText, const kj::HttpHeaders& headers) override {
+    KJ_REQUIRE(statusCode >= 200 && statusCode < 300, "the statusCode must be 2xx for accept");
+
+    auto headersCopy = kj::heap(headers.clone());
+
+    auto pipe = kj::newTwoWayPipe();
+
+    fulfiller->fulfill(kj::HttpClient::ConnectResponse {
+      statusCode,
+      statusText,
+      headersCopy.get(),
+      pipe.ends[0].attach(kj::addRef(*this)),
+    });
+    return kj::mv(pipe.ends[1]);
+  }
+
+  kj::Own<kj::AsyncOutputStream> reject(
+      uint statusCode, kj::StringPtr statusText, const kj::HttpHeaders& headers,
+      kj::Maybe<uint64_t> expectedBodySize) override {
+    KJ_ASSERT(false, "Not testing this branch.");
+    KJ_UNREACHABLE;
+  }
+private:
+  kj::Own<kj::PromiseFulfiller<kj::HttpClient::ConnectResponse>> fulfiller;
+  kj::Promise<void> task = nullptr;
+};
+
+KJ_TEST("CONNECT Service Adapter handles stream closure correctly") {
+  KJ_HTTP_TEST_SETUP_IO;
+
+  auto pipe = KJ_HTTP_TEST_CREATE_2PIPE;
+
+  kj::TimerImpl timer(kj::origin<kj::TimePoint>());
+
+  HttpHeaderTable table;
+  ConnectCloseService service(table);
+  HttpServer server(timer, table, service);
+
+  auto listenTask = server.listenHttp(kj::mv(pipe.ends[0]));
+
+  auto client = newHttpClient(table, *pipe.ends[1]);
+  auto httpService = newHttpService(*client);
+
+  HttpHeaderTable clientHeaders;
+  auto paf = kj::newPromiseAndFulfiller<kj::HttpClient::ConnectResponse>();
+  auto tunnel = kj::refcounted<TestConnectResponseImpl>(kj::mv(paf.fulfiller));
+  // The below promise is fulfilled when the read below finishes successfully.
+  auto readPaf = kj::newPromiseAndFulfiller<void>();
+  httpService->connect("https://example.org"_kj, HttpHeaders(clientHeaders), *tunnel).detach(
+      [](kj::Exception err) { KJ_DBG(err); KJ_FAIL_ASSERT(false); });
+  paf.promise.then([&](HttpClient::ConnectResponse response) -> kj::Promise<void> {
+    KJ_ASSERT(response.statusCode == 200);
+    KJ_ASSERT(response.statusText == "OK"_kj);
+
+    auto& io = KJ_ASSERT_NONNULL(response.connectionOrBody.tryGet<kj::Own<kj::AsyncIoStream>>());
+    char buffer;
+    // Attempt a read. It should disconnect. We are verifying here that the stream is shutdown
+    // correctly, if it doesn't this will block forever (and KJ will give us a
+    // "this thread would hang forever" error message).
+    return io->tryRead(&buffer, 1, 1).then([&](size_t count) {
+      KJ_ASSERT(count == 0, "Expecting the stream to get disconnected.");
+      readPaf.fulfiller->fulfill();
+    }).attach(kj::mv(io), kj::mv(buffer));
```

Comment:
In this case, attaching the buffer does not prolong its lifetime. ;) Maybe the `char buffer` could live at the next scope out, or be heap-allocated?

Reply:
It shouldn't really matter in this case. The callback doesn't read `buffer` anyway, so I'll just remove it from the attach statement.

Reply:
> The callback doesn't read buffer anyway

But the implementation of `tryRead()` could write to it if the test were to fail for some reason. Such failures probably won't ever manifest as a segfault, but to guarantee we see such failures as `"Expecting the stream to get disconnected."` assertion failures I think technically the buffer should be valid for `tryRead()`.